streams/src/test/java/org/apache/kafka/streams/integration/ResetPartitionTimeIntegrationTest.java

```diff
@@ -0,0 +1,279 @@
+            .to(outputRaw, Produced.with(STRING_SERDE, Serdes.Long()));
+
+        final Properties streamsConfig = new Properties();
+        streamsConfig.put(StreamsConfig.DEFAULT_TIMESTAMP_EXTRACTOR_CLASS_CONFIG, MaxTimestampExtractor.class);
+        streamsConfig.put(StreamsConfig.APPLICATION_ID_CONFIG, appId);
+        streamsConfig.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfig.put(StreamsConfig.POLL_MS_CONFIG, Integer.toString(DEFAULT_TIMEOUT));
+        streamsConfig.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, Integer.toString(DEFAULT_TIMEOUT));
+        streamsConfig.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, eosEnabled ? EXACTLY_ONCE : AT_LEAST_ONCE);
+        streamsConfig.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        
+
+        KafkaStreams driver = getStartedStreams(streamsConfig, builder, true);
+        try {
+            // start sending some records to have partition time committed 
+            produceSynchronouslyToPartitionZero(
+                input,
+                asList(
+                    new KeyValueTimestamp<>("k3", "v3", 5000)
+                )
+            );
+            verifyOutput(
+                outputRaw,
+                asList(
+                    new KeyValueTimestamp<>("k3", 1L, 5000)
+                )
+            );
+            assertThat(lastRecordedTimestamp, is(5000L));
+            lastRecordedTimestamp = -1L;
+            Thread.sleep(1000); // wait for commit to finish
+
+            // restart && reset the driver
+            driver.close();
+            assertThat(driver.state(), is(KafkaStreams.State.NOT_RUNNING));
+            driver.cleanUp();
+            driver = new KafkaStreams(builder.build(), streamsConfig);
+            driver.start();
+
+            // resend some records and retrieve the last committed timestamp
+            produceSynchronouslyToPartitionZero(
+                input,
+                asList(
+                    new KeyValueTimestamp<>("k4", "v4", 4998),
+                    new KeyValueTimestamp<>("k5", "v5", 4999)
+                )
+            );
+            verifyOutput(
+                outputRaw,
+                asList(
+                    new KeyValueTimestamp<>("k4", 1L, 4998),
+                    new KeyValueTimestamp<>("k5", 1L, 4999)
+                )
+            );
+            // verify that the lastRecordedTimestamp is 5000
+            assertThat(lastRecordedTimestamp, is(5000L));
+
+            metadataValidator.raiseExceptionIfAny();
+
+        } finally {
+            driver.close();
+            cleanStateAfterTest(CLUSTER, driver);
+        }
+    }
+
+    public static final class MaxTimestampExtractor implements TimestampExtractor {
+        @Override
+        public long extract(final ConsumerRecord<Object, Object> record, final long maxTimestamp) {
+            if (record.timestamp() > maxTimestamp) {
```

Comment:
We should not care about `record.timestamp()` IMHO, but instead use a second variable similar to `lastRecordedTimestamp`: something like `Map<Integer,Long> expectedPartitionTimePerPartition`. This allows us to set an expected partition time per partition and compare it to the passed in `partitionTime` (what you now call `maxTimestamp`)

If passed in partition time does not match expected partition time, we can just throw an `RuntimeException` what will kill KafkaStreams and the test will eventually time out.

Reply:
Oh, the thing is this logic is neccessary, or otherwise the test will crash.

If you were to look closely at ```RecordQueue``` logic, the class updates the partitionTime _after_ TimestampExtractor was called. This is important, because partitionTime is always passed in first. 

That means that ```RecordQueue.UNKNOWN``` is always passed in, and if we always return maxTimestamp without checking if record.timestamp() is greater, than that means -1 will be returned no matter how many records are passed through TimestampExtractor.

Thus, what we do here is stimulate an update to partitionTime _before_ it is actually updated in ```RecordQueue```.

Reply:
> That means that RecordQueue.UNKNOWN is always passed in first when we first start processing

Yes.

> and if we always return maxTimestamp without checking if record.timestamp() is greater, than that means -1 will be returned no matter how many records are passed through TimestampExtractor.

Yes, but you code return `return record.timestamp();` anyway. So for the second call of `extract()` method, `partitionTime` (ie, `maxTimestamp`) gets advanced. For a stop-restart of `KafakStreams`, with this fix on restart `partitionTime` should be `UNKNOWN` any longer, as it's should be initialized from the commit-metadata that is preserved.

Hence, we should see `UNKNOWN` only a single time, and the integration test should verify that we only see on the first start of `KafakStreams` but not for the second start.

> Thus, what we do here is stimulate an update to partitionTime before it is actually updated in RecordQueue.

I cannot follow here. `partitionTime` is tracked internally and it will be updated after TimestampExtractor returns, base on the value that is provided in `return`.

Reply:
Ah, okay. Then I will change that.